File: externals/openmpi-1.6.5/ompi/contrib/vt/vt/tools/opari/lib/transformations.f90
      module xxx
      contains
        subroutine writelog(msg, val)
          character(len=*), intent(in)  :: msg
          integer, optional, intent(in) :: val
          integer, external             :: omp_get_thread_num

          if ( present(val) ) then
            write(*,*) "---", omp_get_thread_num(), ": ", msg, val
          else
            write(*,*) "---", omp_get_thread_num(), ": ", msg
          endif
          return
        end subroutine writelog
      end module xxx

      program trans
      use xxx
      integer, parameter :: iterations = 4
      integer            :: i, k = 0
      integer            :: lck
      logical, external  :: omp_test_lock

      !$omp inst init

      call writelog("sequential", 0);

      ! ---- plain parallel region
      !$omp parallel
        call writelog("parallel", 0);
      !$omp end parallel
 
      call writelog("sequential", 1)

      ! ---- parallel region
      !$omp parallel
        call writelog("parallel", 1)

        !$omp inst begin(worksharing)

        ! ---- worksharing for loop without synchronisation
        !$omp do
        do i=1,iterations
          call writelog("for nowait iteration", i)
        enddo
        !$omp end do nowait

        ! ---- user specified barrier
        !$omp barrier

        ! ---- worksharing for loop with implicit synchronisation
        !$omp do
        do i=1,iterations
          call writelog("for iteration", i)
        enddo
        !$omp end do
 
        ! ---- worksharing tasks without synchronisation
        !$omp sections
          !$omp section
            call writelog("section nowait", 1)
          !$omp section
            call writelog("section nowait", 2)
        !$omp end sections nowait

        ! ---- worksharing tasks with implicit synchronisation
        !$omp sections
          !$omp section
            call writelog("section", 1)
          !$omp section
            call writelog("section", 2)
        !$omp end sections

        !$omp inst end(worksharing)

        !$omp inst begin(synchronisation)

        ! ---- critical section
        !$omp critical
          call writelog("critical\n")
          k = k+1
        !$omp end critical

        ! ---- named critical section
        !$omp critical(kincr)
          call writelog("critical\n")
          k = k+1
        !$omp end critical(kincr)

        ! ---- atomic expression
        !$omp atomic
          k = k+1

        ! ---- update k just once without synchronisation
        !$omp single
          call writelog("single nowait\n")
          k = k+1
        !$omp end single nowait

        ! ---- update k just once with implicit synchronisation
        !$omp single
          call writelog("single\n")
          k = k+1
        !$omp end single

        !$omp master
          call writelog("master\n")
          write(*,*) "k = ", k
          k = 0
        !$omp end master

        !$omp inst end(synchronisation)
      !$omp end parallel

      call writelog("sequential", 2)

      !$omp inst begin(parallelworksharing)

      ! ---- combined parallel worksharing for loop
      !$omp parallel do         &
      !$omp   reduction(+:k)    &
      !$omp   private(i)        &
      !$omp   schedule(dynamic)
      do i=1,iterations
        call writelog("pfor", i)
      enddo
      !$omp end parallel do

      call writelog("sequential", 3)

      ! ---- combined parallel worksharing tasks
      !$omp parallel sections
        !$omp section
          call writelog("psection", 1)
        !$omp section
          call writelog("psection", 2)
      !$omp end parallel sections

      call writelog("sequential", 4)

      !$omp inst end(parallelworksharing)

      ! ---- OpenMP locking API
      !$omp inst begin(locking)


      call omp_init_lock(lck)

      !$omp parallel shared(lck)
        call omp_set_lock(lck)
        call writelog("got lock")
        call omp_unset_lock(lck)

        do while (.not. omp_test_lock(lck))
          call writelog("skipping")
        enddo
        call writelog("working")
        call omp_unset_lock(lck)
      !$omp end parallel

      call omp_destroy_lock(lck)

      !$omp flush(k)
      !$omp inst end(locking)

      end program trans

File: externals/openmpi-1.6.5/ompi/contrib/vt/vt/tools/opari/lib/transformations.cc
#include <iostream>
  using std::cout;
  using std::endl;
#include <iomanip>
  using std::setw;

#include "opari_omp.h"

namespace {
  void log(const char* msg, int val = -1) {
    cout << "--- " << setw(3) << omp_get_thread_num() << ": " << msg;
    if ( val != - 1 ) cout << " " << val;
    cout << endl;
  }
}

const int iterations = 4;

int main() {
  int i, k = 0;

  #pragma omp inst init

  log("sequential", 0);

  // ---- plain parallel region
  #pragma omp parallel
  {
    log("parallel", 0);
  }

  log("sequential", 1);

  // ---- large parallel region
  #pragma omp parallel
  {
    log("parallel", 1);

    #pragma omp inst begin(worksharing)

    // ---- worksharing for loop without synchronisation
    #pragma omp for nowait
    for(i=0; i<iterations; ++i) {
      log("for nowait iteration", i);
    }

    // ---- user specified barrier
    #pragma omp barrier

    // ---- worksharing for loop with implicit synchronisation
    #pragma omp for
    for(i=0; i<iterations; ++i)
      log("for iteration", i);
 
    // ---- worksharing tasks without synchronisation
    #pragma omp sections nowait
    {
      #pragma omp section
        log("section nowait", 1);
      #pragma omp section
      {
        log("section nowait", 2);
      }
    }

    // ---- worksharing tasks with implicit synchronisation
    #pragma omp sections
    {
      #pragma omp section
      {
        log("section", 1);
      }
      #pragma omp section
        log("section", 2);
    }

    #pragma omp inst end(worksharing)

    #pragma omp inst begin(synchronisation)

    // ---- critical section
    #pragma omp critical
    {
      log("critical\n");
      k += 1;
    }

    // ---- named critical section
    #pragma omp critical(kincr)
    {
      log("critical\n");
      k += 1;
    }

    // ---- atomic expression
    #pragma omp atomic
      k += 1;

    // ---- update k just once without synchronisation
    #pragma omp single nowait
    {
      log("single nowait\n");
      k += 1;
    }

    // ---- update k just once with implicit synchronisation
    #pragma omp single
    {
      log("single\n");
      k += 1;
    }

    #pragma omp master
    {
      log("master\n");
      cout << "k = " << k << endl;
      k = 0;
    }

    #pragma omp inst end(synchronisation)
  } // end parallel

  log("sequential", 2);

  #pragma omp inst begin(parallelworksharing)

  // ---- combined parallel worksharing for loop
  #pragma omp parallel for \
          reduction(+:k)  \
          private(i)       \
          schedule(dynamic)
  for(i=0; i<iterations; ++i) {
    log("pfor", i);
  }

  log("sequential", 3);

  // ---- combined parallel worksharing tasks
  #pragma omp parallel sections
  {
    #pragma omp section
      log("psection", 1);
    #pragma omp section
      log("psection", 2);
  }

  log("sequential", 4);

  #pragma omp inst end(parallelworksharing)

  
  // ---- OpenMP locking API
  #pragma omp inst begin(locking)

  omp_lock_t lck;

  omp_init_lock(&lck);

  #pragma omp parallel shared(lck)
  {
    omp_set_lock(&lck);
    log("got lock");
    omp_unset_lock(&lck);

    while (! omp_test_lock(&lck)) {
      log("skipping");
    }
    log("working");
    omp_unset_lock(&lck);
  }

  omp_destroy_lock(&lck);

  #pragma omp flush(k)

  #pragma omp inst end(locking)
}
